test: make temp path customizable

jbergstroem opened the pull request:
In CI we previously passed `NODE_COMMON_PIPE` to the test runner to avoid long filenames on Windows. Avoid long paths by changing the temporary directory instead. This also allows us to run test suites in parallel since `NODE_COMMON_PIPE` otherwise would have been used from multiple tests.

/R=@nodejs/build, @Trott ?

jbergstroem commented:
CI (without passing `NODE_TMP_DIR`): https://ci.nodejs.org/job/node-test-commit/804/

rvagg commented on test/common.js:
```diff
@@ -6,6 +6,8 @@ var assert = require('assert');
 var os = require('os');
 var child_process = require('child_process');
 
+const testRoot = process.env.NODE_TEST_DIR || path.dirname(__filename);
```
`__dirname` perhaps?

rvagg commented on test/common.js:
```diff
@@ -6,6 +6,8 @@ var assert = require('assert');
 
+const testRoot = process.env.NODE_TEST_DIR || path.dirname(__filename);
+
 exports.testDir = path.dirname(__filename);
```
`__dirname`

jbergstroem replied:
Ok. Rationale behind `__filename` was previous default.

rvagg commented:
I'm still catching up but did `NODE_COMMON_PIPE` get removed from all the tests already?

jbergstroem commented:
AFAIK it was never in tests, just passed to `exports.PIPE`.

Trott commented:
I'll leave it to someone more in tune with the entirety of the build process to `LGTM` or not, but I'm +1. The `common.PIPE` improvements that build on this to enable parallel testing will be in a subsequent PR?

jbergstroem commented:
Prior to removal, if `NODE_COMMON_PIPE` was unset `exports.PIPE` would be stored in the temporary directory which works fine in parallel runs since the thread id is appended to `common.tmpDir`. The change I'm introducing is the ability to change where `tmpDir` lives.

jasnell commented:
@jbergstroem ... what's the status on this? Looks like it needs to be rebased at a minimum

jbergstroem commented:
@jasnell waiting for feedback/review. I'll rebase.

jbergstroem commented:
@Trott to answer your question regarding "enabling" parallel builds -- that would be something we'd most likely be changing in the jenkins setup.

joaocgreis commented:
LGTM

Can you ensure that `testRoot` exists? Since this is now a variable, it can be useful to set it to a directory that does not exist yet.

To run parallel tests on Windows, the pipe name should also depend on the thread number. What about adding something like lines 64 and 66 after 139?

jbergstroem commented:
@joaocgreis does the pipe really have to depend on a thread number since it already lives in its own thread folder (`tmp.$thread`)? 

Regarding creating the directory, that's already part of `refreshTmpDir`.

joaocgreis commented:
@jbergstroem the pipe is defined differently in windows, it is a global name, not a file in the test folder (as created in [line 139](https://github.com/nodejs/node/blob/3f1add0b24c5ff849d47d07c6029153f1b6ac754/test/common.js#L139)).

The `tmp.$thread` directories are created by `refreshTmpDir`, but `testRoot` is not. Currently this is not a problem since the test directory always exists, but if you set `NODE_TEST_DIR` to a directory that does not exit, tests fail. There could be a `mkdir(testRoot)` or just a descriptive exception if it does not exist, but on the other hand the error is pretty clear so feel free to ignore this suggestion.

jbergstroem commented:
Rebased and fixed an issue with two tests that assumed a specific folder structure. CI (which again shouldn't branch into this): https://ci.nodejs.org/job/node-test-commit/1324/

jbergstroem commented:
Oh, and @joaocgreis -- I now create the dir through refreshTmp.

Trott commented on test/common.js:
```diff
@@ -64,18 +66,19 @@ function rmdirSync(p, originalEr) {
 }
 
 exports.refreshTmpDir = function() {
+  if (!fs.existsSync(testRoot)) {
```
[`existsSync()` is deprecated](https://nodejs.org/api/fs.html#fs_fs_existssync_path) so maybe use one of the suggested alternatives?

jbergstroem commented:
ping!

rvagg commented on test/common.js:
```diff
@@ -155,21 +162,13 @@ Object.defineProperty(exports, 'hasFipsCrypto', {
 if (exports.isWindows) {
   exports.PIPE = '\\\\.\\pipe\\libuv-test';
+  if (process.env.TEST_THREAD_ID) {
+    exports.PIPE = +'.' + process.env.TEST_THREAD_ID;
```
needs space between `+` and `'.'`

joaocgreis replied:
should be `+=`

rvagg commented on test/common.js:
```diff
@@ -64,18 +66,23 @@ function rmdirSync(p, originalEr) {
+    fs.accessSync(testRoot, fs.W_OK);
+  } catch (e) {
+    if (e.code === 'ENOENT') {
+      fs.mkdirSync(testRoot);
```
I'm not convinced about this, it really needs a mkdirp because there's zero guarantee that the parent will exist, perhaps the user should get the error if they have tried to supply a non-existent directory?

jbergstroem replied:
Yeah, that was my original approach (bail and let the user create it) by @joaocgreis suggested I'd try and create it. Should I invoke `mkdirp` instead of `mkdirSync`?

joaocgreis commented:
@jbergstroem wouldn't checking the NODE_TEST_DIR directly in the test runner be better?

jbergstroem commented:
@joaocgreis checking in the runner now. Also, created an option called `--test-dir` supplementing `NODE_TEST_DIR`.

jbergstroem commented:
CI (should be unaffected): https://ci.nodejs.org/job/node-test-commit/1560/

jbergstroem commented:
Test CI run with parallel tests (yay): https://ci.nodejs.org/job/node-test-commit/1561/

jbergstroem commented:
@Trott a few failing tests across platforms as a result from the parallelism. I'll back `-J` out for merging this issue and will open a new PR to try and flesh that out.

Trott commented:
Looking at `test-net-pipe-connect-errors.js` on the Raspberry Pi devices, the assertion is firing because the error code received is `ENOENT` when `ENOTSOCK` or `ECONNREFUSED` are expected.

Trott commented:
It sure looks like the problem with `test-net-pipe-connect-errors.js` on the Raspberry Pi devices is that the resulting path exceeds the 108 character POSIX limit. Is that plausible? 

At this particular point in the test, it isn't using `common.PIPE` but rather `common.PIPE` plus a `.txt` extension. So maybe the reason it's not failing in other tests is because it hovers right around 105 or 106 characters and only exceeds 108 when `.txt` is appended to it?

I created a 115-character path and used it as the first argument to `net.createConnection()`. An error event fired and the code was `ENOENT`. So that would be consistent with what's going on in this test.

jbergstroem commented:
@Trott: That's good because we can now change temp path to `~/tmp` or similar.

joaocgreis commented:
LGTM without the `-J`

mhdawson commented:
Just back from holidays and I've noticed this affected our internal builds.  Its an easy thing to fix but I wonder if there is any need to give people who may be running the tests in their own environments a heads up.  One idea might be a mailing list for those who are interested in changes that affect how you have to build/test.

The other thing I saw was this failure

not ok 254 test-fs-link.js
# /jenkins-workspace/slave-pales/workspace/convergance-test-aix/arch/ppc/build/build/gcc48/gcc48/os/linux/node/test/parallel/test-fs-link.js:14
#   if (err) throw err;
#            ^
# 
# Error: EXDEV: cross-device link not permitted, link '/jenkins-workspace/slave-pales/workspace/convergance-test-aix/arch/ppc/build/build/gcc48/gcc48/os/linux/node/test/fixtures/cycles/root.js' -> '/tmp/NodeTestDirMw6MwJyDSI/tmp.0/link1.js'
#     at Error (native)
  ---


So we at least should document that the test directory needs to be on the same device as where node has been built.  Now I just have to figure out how to chose a short enough one that is on the save device

jbergstroem commented:
@mhdawson what issues did you run into more specifically? Path issues as a result of the path to pipe was getting too long?

Regarding cross-linking; you'll see a few of those. I get them locally when working between mounted volumes because you most often can symlink between mounts. I suggest you set the tempdir to something that lives on the same volume.

mhdawson commented:
@jbergstroem the issue was the failing test above.  Under jenkins I think what I can get is $WORKSPACE but that is already too deep so I ended up having to do NODE_TEST_TMP=`readlink -f $WORKSPACE/../../../../../../../..` to get the directory to use, but I'm not too happy with that.

mhdawson commented:
My comment about a heads up was not info about specific issues but more that something is going to change that could break your builds.

jasnell commented:
Landed in v4.x-staging in d33279d and b3079ef
